gpu4pyscf/tdscf/ris.py

```diff
@@ -0,0 +1,1653 @@
+                int3c_slice_blk = cart2sph(int3c_slice_blk, axis=2, ang=li)
+
+
+            if omega and omega != 0:
+                int3c_slice_blk_omega = get_int3c2e_slice(intopt, cp_ij_id, cp_kl_id, omega=omega)
+
+                if not mol.cart:
+                    int3c_slice_blk_omega = cart2sph(int3c_slice_blk_omega, axis=1, ang=lj)
+                    int3c_slice_blk_omega = cart2sph(int3c_slice_blk_omega, axis=2, ang=li)
+                int3c_slice_blk = alpha * int3c_slice_blk +  beta * int3c_slice_blk_omega
+
+            int3c_slice_blk = cp.asarray(int3c_slice_blk, dtype=cp.float32, order='C')
+            i0, i1 = intopt.ao_loc[cpi], intopt.ao_loc[cpi+1]
+            j0, j1 = intopt.ao_loc[cpj], intopt.ao_loc[cpj+1]  
+            
+            assert int3c_slice[:,j0:j1, i0:i1].shape == int3c_slice_blk.shape
+            int3c_slice[:,j0:j1, i0:i1] = int3c_slice_blk
+
+        if aosym:
+            row, col = cp.tril_indices(nao)
+            int3c_slice[:, row, col] = int3c_slice[:, col, row]
+
+
+        unsorted_ao_index = cp.argsort(intopt._ao_idx)
+        int3c_slice = int3c_slice[:, unsorted_ao_index, :]
+        int3c_slice = int3c_slice[:, :, unsorted_ao_index]
+
+        if 'J' in calc:
+            Ppq[k0:k1,:,:] = get_PuvCupCvq_to_Ppq(int3c_slice,C_p,C_q)
+
+        if 'K' in calc:
+
+            Ppp[k0:k1,:,:] = get_PuvCupCvq_to_Ppq(int3c_slice,C_p,C_p)
+            Pqq[k0:k1,:,:] = get_PuvCupCvq_to_Ppq(int3c_slice,C_q,C_q)
+
+
+    unsorted_aux_ao_index = cp.argsort(intopt._aux_ao_idx)
+
+
+    # DEBUG = False
+    # if DEBUG:
+    #     eri_3c2e = get_int3c2e(mol, auxmol, omega=0)
+    #     if omega and omega != 0:
+    #         eri_3c2e_erf = get_int3c2e(mol, auxmol, omega=omega)
+    #         eri_3c2e = alpha * eri_3c2e + beta * eri_3c2e_erf
+    #     tmp = cp.einsum('Puv,up->Ppv', eri_3c2e, C_p)
+    #     Ppq = cp.einsum('Ppv,vq->Ppq', tmp, C_q)
+
+        
+    if calc == 'J':
+        Tpq = get_Ppq_to_Tpq(Ppq[unsorted_aux_ao_index,:,:], lower_inv_eri2c)
+        return Tpq
+
+    if calc == 'K':
+        Tpp = get_Ppq_to_Tpq(Ppp[unsorted_aux_ao_index,:,:], lower_inv_eri2c)
+        Tqq = get_Ppq_to_Tpq(Pqq[unsorted_aux_ao_index,:,:], lower_inv_eri2c)
+        return Tpp, Tqq
+
+    if calc == 'JK':
+        Tpq = get_Ppq_to_Tpq(Ppq[unsorted_aux_ao_index,:,:], lower_inv_eri2c)
+        Tpp = get_Ppq_to_Tpq(Ppp[unsorted_aux_ao_index,:,:], lower_inv_eri2c)
+        Tqq = get_Ppq_to_Tpq(Pqq[unsorted_aux_ao_index,:,:], lower_inv_eri2c)
+        return Tpq, Tpp, Tqq
+
+def get_eri2c_inv_lower(auxmol, omega=0, alpha=None, beta=None):
+
+    eri2c = auxmol.intor('int2c2e')
+
+    if omega and omega != 0:
+
+        auxmol.set_range_coulomb(omega)
+        auxmol.build()
+        eri2c_erf = auxmol.intor('int2c2e')
+        eri2c = alpha * eri2c + beta * eri2c_erf
+
+    eri2c = cp.asarray(eri2c, dtype=cp.float64, order='C')
+    lower_inv_eri2c = math_helper.matrix_power(eri2c,-0.5,epsilon=1e-8)

```

Comment:
1. Maybe first call the cholesky factorization with solve(cholesky, eri3c) . If that does not work, turning to the matrix^{-1/2} with dot(inv, eri3c). cholesky factorization is potentially more accurate.
2.  the magic number 1e-8 can be a constant variable defined at module level